lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -240,7 +243,8 @@ let next dis s =
     | r :: roots when not(Memory.contains s.base r) ->
       loop {s with roots}
     | r :: roots when Span.mem s.visited r ->
-      loop {s with roots}
+      if Set.mem s.insns r then loop {s with roots}
+      else loop {s with roots; inits = Set.remove s.inits r}
```

Comment:
This code will remove an invalid root only in case if we have discovered that it is invalid before we extract it from the worklist. Since the worklist is an unordered sequence, this won't work in general.